File: backend/atlas/keys/guardian_forensics.py
"""Guardian forensic access — a trusted contact can collect your panic forensics ONLY if BOTH of you
agreed. Consent is enforced by arithmetic, not policy.

The forensic capture is sealed under a `forensic_key`, which is Shamir 2-of-2 split into:
  * a CONTACT share — handed to the trusted contact when they ACCEPT the guardian role (their consent).
  * an OWNER share — kept by you and RELEASED ONLY when panic fires (your consent, in the moment).

Neither share alone opens anything. So the contact can reconstruct `forensic_key` and read your
capture only when they hold their share (they agreed) AND you released yours (you triggered panic) —
"both parties agree" as a 2-of-2, not a promise. A stolen contact-share alone is inert; a foreign
owner's share combined with this contact's share reconstructs a WRONG key that fails to open the
capture (AEAD rejects it) — shares are per-relationship, so cross-pairing collects nothing.

Reference of record. Composes atlas.crypto.shamir + AES-GCM; no new primitive.
"""
from __future__ import annotations

from dataclasses import dataclass
from typing import Tuple

from ..crypto import shamir
from ..crypto.primitives import aead_decrypt, aead_encrypt, random_bytes


class GuardianError(Exception):
    """Base — fail-closed."""


class ConsentIncomplete(GuardianError):
    """Collection attempted without BOTH the contact share (accepted) and the owner share (released)."""


@dataclass(frozen=True)
class GuardianGrant:
    """What each side ends up holding after a mutual guardian setup."""
    contact_share: shamir.Share    # given to the trusted contact when they accept (their standing consent)
    owner_share: shamir.Share      # kept by the owner; released to the contact only on panic


def setup_guardian(forensic_key: bytes) -> GuardianGrant:
    """Split the forensic key 2-of-2 at guardian setup. Call when the owner designates a contact AND
    the contact accepts — the contact keeps `contact_share`, the owner keeps `owner_share`."""
    if len(forensic_key) < 16:
        raise ValueError("forensic_key must be >= 16 bytes")
    a, b = shamir.split(forensic_key, n=2, k=2)
    return GuardianGrant(contact_share=a, owner_share=b)


def collect_forensic_key(contact_share: shamir.Share, owner_share: shamir.Share) -> bytes:
    """Reconstruct the forensic key — needs BOTH shares (contact accepted + owner released on panic)."""
    try:
        return shamir.combine([contact_share, owner_share])
    except Exception as e:  # a single share, or a foreign owner's share, yields the wrong/no key
        raise ConsentIncomplete("need BOTH the contact's share and the owner's panic-released share") from e


# Convenience: seal / open the capture itself under the forensic key.
def seal_forensic(forensic_key: bytes, data: bytes, *, aad: bytes = b"atlas/guardian/forensic") -> bytes:
    return aead_encrypt(forensic_key, data, aad=aad)


def open_forensic(forensic_key: bytes, blob: bytes, *, aad: bytes = b"atlas/guardian/forensic") -> bytes:
    return aead_decrypt(forensic_key, blob, aad=aad)


def new_forensic_key() -> bytes:
    return random_bytes(32)

File: ios/AtlasCore/Sources/AtlasCore/Keys/GuardianForensics.swift
import Foundation

/// Guardian forensic access — a trusted contact can collect your panic forensics ONLY if BOTH of you
/// agreed. Consent is arithmetic, not policy. Parity with `atlas/keys/guardian_forensics.py`.
///
/// The forensic capture is sealed under a `forensicKey`, Shamir 2-of-2 split into:
///   * a CONTACT share — handed to the trusted contact when they ACCEPT the guardian role (consent).
///   * an OWNER share — kept by you, RELEASED ONLY when panic fires (your consent, in the moment).
///
/// Neither share alone opens anything. A stolen contact-share is inert; a foreign owner's share
/// paired with this contact's share reconstructs a WRONG key that fails to open the capture (AEAD
/// rejects it). Shares are per-relationship, so cross-pairing collects nothing.
public enum GuardianForensics {

    public enum GuardianError: Error, Equatable {
        case forensicKeyTooShort
        case consentIncomplete   // missing the contact share (accepted) or the owner share (released)
    }

    /// What each side ends up holding after a mutual guardian setup.
    public struct GuardianGrant: Equatable {
        public let contactShare: Shamir.Share   // given to the contact on accept (standing consent)
        public let ownerShare: Shamir.Share     // kept by the owner; released to the contact on panic
        public init(contactShare: Shamir.Share, ownerShare: Shamir.Share) {
            self.contactShare = contactShare
            self.ownerShare = ownerShare
        }
    }

    /// Split the forensic key 2-of-2 at guardian setup (owner designates + contact accepts).
    public static func setupGuardian(forensicKey: Data) throws -> GuardianGrant {
        guard forensicKey.count >= 16 else { throw GuardianError.forensicKeyTooShort }
        let shares = Shamir.split(forensicKey, n: 2, k: 2)
        return GuardianGrant(contactShare: shares[0], ownerShare: shares[1])
    }

    /// Reconstruct the forensic key — needs BOTH shares (contact accepted + owner released on panic).
    public static func collectForensicKey(contactShare: Shamir.Share, ownerShare: Shamir.Share) throws -> Data {
        guard contactShare.index != ownerShare.index else { throw GuardianError.consentIncomplete }
        return Shamir.combine([contactShare, ownerShare])
    }

    public static func sealForensic(forensicKey: Data, data: Data,
                                    aad: Data = Data("atlas/guardian/forensic".utf8)) throws -> Data {
        try Primitives.aeadEncrypt(key: forensicKey, plaintext: data, aad: aad)
    }

    public static func openForensic(forensicKey: Data, blob: Data,
                                    aad: Data = Data("atlas/guardian/forensic".utf8)) throws -> Data {
        try Primitives.aeadDecrypt(key: forensicKey, blob: blob, aad: aad)
    }

    public static func newForensicKey() -> Data { Primitives.randomBytes(32) }
}
